ui: statement details page + snazzy up the queries page

couchand opened the pull request:
This adds a statement details page, which is linked to from the queries list page, as well as adding some enhancements to that listing.

Statements list page:
<img width="1258" alt="screen shot 2018-06-19 at 1 22 27 pm" src="https://user-images.githubusercontent.com/793969/41613687-48aad60a-73c4-11e8-8254-3fbe8c7adaca.png">


Statement details page:
<img width="1242" alt="screen shot 2018-06-19 at 1 22 16 pm" src="https://user-images.githubusercontent.com/793969/41613698-4f72420c-73c4-11e8-9039-272416b4f2af.png">


A visual guide to reading the latency chart:
<img width="163" alt="statement-stats" src="https://user-images.githubusercontent.com/793969/41618187-7402202c-73d0-11e8-9673-feaa4d557b33.png">


Fixes #26525

couchand committed:
server, ui: add time of last reset of statement stats to queries page

This adds the last reset timestamp to the API call for statement
stats, so that the UI can show how much data is available.

Release note (admin ui change): Added to the queries page the
time that the statement stats were last reset.

couchand committed:
ui: start to summarize statement

Provide a short summary of the statement in the table itself, and
provide the entire text of the query as a tooltip.  It will also
be on the full page for the query.

Release note: None

couchand committed:
ui: add page with statement details

This change creates a new page with the detailed statistics about
statement execution.  It will continue to get more data added and the
design refined.

Release note (admin ui change): Add more details about statement execution.

couchand committed:
ui: add bar charts for columns

This is a first pass at bar charts for the statement statistics.

Release note (admin ui change): Visualize the statement statistics
on the queries page with bar charts.

couchand committed:
ui: make statement stats bars stacked

This changes the statement stats bars to be stacked bars of multiple
statistics: the count stacks retries to the right of first tries, and
the latency column stacks each phase latency.

Release note: None

couchand committed:
ui: add bar chart to statement details page

This adds the same bar charts previously added to the queries page
to the statement details page, with tables underneath each to break
out the detailed numbers.

Release note: None

couchand committed:
ui: make typescript happy

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26711)
<!-- Reviewable:end -->

vilterp commented on pkg/ui/src/views/queries/statementDetails.tsx:
```diff
@@ -0,0 +1,209 @@
+    return (
+      <div>
+        <Helmet>
+          <title>{`Details | Statements`}</title>
```
wai not `<title>Details | Statements</title>`?

couchand replied:
Done.

vilterp commented on pkg/ui/src/views/queries/statementDetails.tsx:
```diff
@@ -0,0 +1,209 @@
+
+  render() {
+    if (!this.props.statement) {
+      return "loading...";
```
use `<Loading>`?

couchand replied:
Ha! Done.

vilterp reviewed (commented):
This is hawt. Can't wait to get the plans in there.

Left a couple of minor comments.

(screenshots taken out as PR desc now has them)

vilterp commented on pkg/ui/src/views/queries/index.tsx:
```diff
@@ -30,29 +34,57 @@ interface QueriesPageState {
+      sort: (query) => FixLong(query.stats.count).toInt(),
+    },
+    {
+      title: "Avg Rows",
```
maybe change this and "latency" below to "mean", since you seem to be using that everywhere else

couchand replied:
Done.

vilterp commented on pkg/ui/src/views/queries/statementDetails.tsx:
```diff
@@ -0,0 +1,209 @@
+          <title>{`Details | Statements`}</title>
+        </Helmet>
+        <section className="section">
+          <h2>Statement Details</h2>
```
Hate to bikeshed on terminology, but maybe rename the main page to "statements" too? Meh, I dunno.

Also, maybe "Statement Fingerprint Details"? Probably want to reinforce that this is a set of queries with the same fingerprint, not a single query.

Amruta-Ranade replied:
"Statements" is definitely growing on me. Seems unambiguous and non-confusing of all options.

couchand replied:
Done.

vilterp commented:
![image](https://user-images.githubusercontent.com/7341/41382644-146207fc-6f3b-11e8-817c-0c943266d783.png)

Small bug: certain fingerprints don't syntax highlight when I go to them. Can't tell what the pattern is.

mjibson commented on pkg/ui/src/views/queries/statementDetails.tsx:
```diff
@@ -0,0 +1,209 @@
+          <tr className="numeric-stats-table__row--header">
+            <th className="numeric-stats-table__cell" />
+            <th className="numeric-stats-table__cell">Mean</th>
+            <th className="numeric-stats-table__cell">Variance</th>
```
Maybe this has already been designed or bikeshed elsewhere, but was there a reason to choose variance over std deviation? As a human looking at things like this I'm used to seeing and interpreting std deviation. If this stays as variance then I have to square root it in my head to get a gut feeling about what the number means.

vilterp replied:
I think we just copied it over from what was in `crdb_internal.node_statement_statistics`; there hasn't been much of a discussion about it. Could easily square root to get stddev if that is easier to interpret.

mjibson replied:
That'd be preferrable. Variance is easier to math. Stddev is easier for humans.

couchand replied:
Done.

vilterp commented on pkg/ui/src/views/queries/statementDetails.tsx:
```diff
@@ -0,0 +1,209 @@
+          <h2>Statement Details</h2>
+          <div className="content l-columns">
+            <div className="l-columns__left">
+              <SqlBox value={ this.props.statement.key.query } />
```
design tweak: until we figure out how to use @mjibson's sqlfmt to pretty print queries, I suggest applying this diff to `SqlBox` to just wrap the query text:

```diff
diff --git a/pkg/ui/src/views/shared/components/sql/sqlhighlight.styl b/pkg/ui/src/views/shared/components/sql/sqlhighlight.styl
index 3daaf9fdf..b3d15397b 100644
--- a/pkg/ui/src/views/shared/components/sql/sqlhighlight.styl
+++ b/pkg/ui/src/views/shared/components/sql/sqlhighlight.styl
@@ -5,6 +5,7 @@
   padding 20px
   width 100%
   margin-bottom 10px
+  white-space pre-wrap
 
 .sql-highlight.hljs
     background-color $headings-color
```

couchand replied:
Done.

couchand committed:
ui: move queries page to statements

Release note (admin ui change): Renames the queries page to
the more general statements page.

couchand committed:
ui: use stddev on statement details page

Standard deviation is easier for humans.  For one thing, its

couchand committed:
ui: use loading component for statement details

Release note: None

couchand committed:
ui: some statement details cleanup

Release note: None

couchand committed:
ui: add total time to statements page

Release note (admin ui change): Add a column to the Statements page
listing the total time spent servicing each statement.  This provides
a good general metric for the load a statement is putting on the
cluster.

couchand committed:
ui: add app to statements page

Release note (admin ui change): Add app name to the Statements page
and Statement Details page.

couchand committed:
ui: link app to filtered statements view

Release note (admin ui change): Update the app name on the Statements
page to link to a filtered view of that app's statements.

couchand commented on pkg/ui/src/views/queries/index.tsx:
```diff
@@ -85,19 +117,20 @@ class QueriesPage extends React.Component<QueriesPageProps, QueriesPageState> {
         <span className="queries-screen__last-hour-note">
           {queries.length} query fingerprints.
-          Query history is only maintained for the past hour.
+          Query history is only maintained for about an hour.
```
Done.

couchand commented on pkg/server/serverpb/admin.proto:
```diff
@@ -553,6 +553,9 @@ message QueriesRequest {
   repeated cockroach.sql.CollectedStatementStatistics queries = 1 [(gogoproto.nullable) = false];
+   // Timestamp of the last stats reset.
+   google.protobuf.Timestamp last_reset = 3
+       [ (gogoproto.nullable) = false, (gogoproto.stdtime) = true ];
```
Done.

couchand commented:
@vilterp I made #26826 to track your highlighting issue.  I'd seen that before and just said to myself "it will be gone when we get rid of highlight.js..."

vilterp commented on pkg/ui/src/views/statements/statementDetails.tsx:
```diff
@@ -0,0 +1,252 @@
+              <tbody>
+                <tr className="numeric-stats-table__row--body">
+                  <th className="numeric-stats-table__cell" style={{ textAlign: "left" }}>App</th>
+                  <td className="numeric-stats-table__cell" style={{ textAlign: "right" }}>{ this.props.statement.key.app }</td>
```
these are hashed server-side before they're returned: https://github.com/cockroachdb/cockroach/blob/dc15c475de05cfde4dabda55c6823d358b5db585/pkg/sql/app_stats.go#L322

![image](https://user-images.githubusercontent.com/7341/41608120-b6ae0302-73b5-11e8-8b8b-0c4927e3b05d.png)

I agree that this is a valuable column to have, but suggest leaving it out and figuring out a way to get the unhashed name in a later PR. (Plus, a widget letting you see and manipulate the filters would be good)

vilterp commented on pkg/ui/src/views/statements/statementsPage.tsx:
```diff
@@ -0,0 +1,215 @@
+    },
+    {
+      title: "Time",
+      cell: (query) => Duration(calculateCumulativeTime(query) * 1e9),
```
Love this column, but seems weird for it to not have a bar chart like the others. I know we're running out of horizontal space; may have to do something clever down the road to regain space like putting stat numbers on top of bars instead of next to them, etc.

vilterp reviewed (approved):
Nice work. I say let's remove the (hashed) app names for now and ship it.

couchand committed:
ui: breakdown bar chart into table rows

This moves the bar chart on the Statement Details page into the various rows
of the table below, making it easier to see the impact of each latency phase
or execution type.

Release note (admin ui change): Improve the bar chart breakdowns on the
Statement Details page.

couchand committed:
ui: approximate counts in statements table

When the counts are large, the exact value isn't needed.  This just formats
the values as, e.g., 115k, 34m, or 9b when they get big enough.

Release note (admin ui change): Update Statements table to only show
approximate query & row counts for large numbers.

couchand committed:
ui: move statement key to right column

The statement key details were a bit out of place in the main column.  This
moves them under the headline stats in the right column of the Statement
Details page.

Release note: None

couchand committed:
ui: fix standard deviation charting

There was an overlap bug in the rendering of the standard deviation
spread on the Statement Details page.  This corrects it.

Release note: None

couchand committed:
ui: use approximify on statement details headline stats

Release note: None

vilterp commented:
Looks like you might have forgotten to check in `app_stats.ts`; getting a build error…

couchand committed:
sql: don't hash app name if statement isn't being scrubbed

This allows us to show the unhashed name on the Statement Details page.

Release note (admin ui change): Show plain text application name
instead of hashed value on Statement Details page.

couchand committed:
ui: remove app column on Statements page

It does not seem to be bearing the weight.

Release note: None

couchand committed:
ui: rename app_stats to appStats

Release note: None

couchand committed:
ui: fix links to statements page on debug page

Release note: None

couchand commented:
bors r+

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26711)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/admin.go, line 1509 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/26711#-LPEU1wsEw5hh51-lFSa:-LPEU1wsEw5hh51-lFSb:b-cf0379) ([raw file](https://github.com/cockroachdb/cockroach/blob/68bd52975c6e252af9385b5d8958a803dc1e2645/pkg/server/admin.go#L1509)):*
> ```Go
> ) (*serverpb.QueriesResponse, error) {
> 	stmtStats := s.server.pgServer.SQLServer.GetUnscrubbedStmtStats()
> 	lastReset := s.server.pgServer.SQLServer.GetStmtStatsLastReset()
> ```

You're retrieving `stmtStats` and `lastReset` in two separate calls here which means that the values might not be consistent. Is that a problem?


<!-- Sent from Reviewable.io -->

couchand commented:
> You're retrieving stmtStats and lastReset in two separate calls here which means that the values might not be consistent. Is that a problem?

Thanks for pointing that out, @petermattis, it's certainly not ideal.  Also, the data for last reset is a little suspicious anyway, at least on multiple-node clusters, see: https://github.com/cockroachdb/cockroach/blob/9104133b280164c2bbee600cd537cf3d4e9b3ddb/pkg/server/statements.go#L74

I'm inclined to not worry too much, since I expect to be removing this well before the next release, and it's probably not worth backporting a fix for this marginal issue.  Does that seem right?
